File: 백준/Gold/9663. N－Queen/N－Queen.js
const input = +require("fs").readFileSync("/dev/stdin").toString().trim();

//열
const visit1 = new Array(input).fill(false);
//대각
const visit2 = new Array(input * 2 - 1).fill(false);
//대각
const visit3 = new Array(input * 2 - 1).fill(false);

let cnt = 0;

function dfs(depth) {
  if (depth === input) {
    cnt++;
    return;
  }

  for (let i = 0; i < input; i++) {
    if (visit1[i] || visit2[i + depth] || visit3[i - depth + input - 1])
      continue;
    visit1[i] = true;
    visit2[i + depth] = true;
    visit3[i - depth + input - 1] = true;
    dfs(depth + 1);
    visit1[i] = false;
    visit2[i + depth] = false;
    visit3[i - depth + input - 1] = false;
  }
}

dfs(0);
console.log(cnt);

File: 백준/Gold/9663. N－Queen/N－Queen.py
N = int(input())
visited1 = [0] * N
visited2 = [0] * (2 * N - 1)
visited3 = [0] * (2 * N - 1)
count = 0


def dfs(depth):
    global count
    if depth == N:
        count += 1
        return

    for i in range(N):
        if not visited1[i] and not visited2[i+depth] and not visited3[i-depth+N-1]:
            visited1[i] = 1
            visited2[i + depth] = 1
            visited3[i-depth+N-1] = 1
            dfs(depth + 1)
            visited1[i] = 0
            visited2[i + depth] = 0
            visited3[i-depth+N-1] = 0


dfs(0)
print(count)